[BACKEND] Implement generic swizzling when lowering `convert_layout`

lezcano opened the pull request:
We implement a generic swizzling algorithm by @apgoucher that, given two linear layouts, finds the optimal shared memory layout that maximises read/write vectorisation and, provided that, minimises bank conflicts.

We also implement an algorithm to find the minimum tile size necessary to perform the `convert_layout` given the restrictions above, and we use it to perform the `convert_layout` iteratively.

This PR does not yet implement a lowering to ldmatrix/stmatrix, we'll do that in a future PR.

lezcano commented on lib/Tools/GenericSwizzling.cpp:
```diff
@@ -0,0 +1,327 @@
+                               smem.getOutDims(),
+                               /*requireSurjective=*/true);
+  return smemReps;
+  // if (reps.size() == 0) {
```
I wrote this once I was figuring out the algorithm. I was able to finally implement it much more cleanly via the `regPermForDivide` tho. Will delete this commented code.

Jokeren commented on lib/Tools/GenericSwizzling.cpp:
```diff
@@ -0,0 +1,327 @@
+  assert((bitwidth > 0) && llvm::isPowerOf2_32(bitwidth) &&
+         "bitwidth must be a power of two");
+
+  const int32_t rank = src.getTotalOutDimSizeLog2();
```
outSize or size might be more intuitive

lezcano replied:
changed it to dim as it is the dimension of the vector space we are working with.

Jokeren commented on lib/Tools/GenericSwizzling.cpp:
```diff
@@ -0,0 +1,327 @@
+  auto logBankConflicts = std::min<int32_t>(
+      std::max<int32_t>(0, lenSegment - A.size() - segment.size()), A.size());
+  // Conflict-free
+  for (int i = logBankConflicts; i < A.size(); ++i)
```
This `^` operator here isn't clear to me, but we can chat offline

lezcano replied:
this part is in the explanation of the algorithm in the paper, but yes, I agree that it is not super intuitive the first time you see it.

Jokeren replied:
Oh I reminded now this is the union of the two subspaces

lezcano replied:
yep! This code finds the largest subspace that's not in the union of the two given ones.

lezcano commented on test/Analysis/test-allocation.mlir:
```diff
@@ -68,7 +68,7 @@ tt.func @matmul_loop(%lb : index, %ub : index, %step : index, %A : !tt.ptr<f16>,
 // Shared memory is available after a tensor's liveness range ends
 // expected-remark @below {{reusable}}
-// expected-remark @below {{size = 4608}}
+// expected-remark @below {{size = 8192}}
```
these often come from being able to vectorise more than before (and as such, not being abl eto do so many reps).

Jokeren replied:
So you haven't seen any cases in internal benchmarks that either slow down or out of shared memory?

lezcano commented on include/triton/Tools/GenericSwizzling.h:
```diff
@@ -0,0 +1,17 @@
+#include "llvm/ADT/ArrayRef.h"
```
oooops, thank you for catching

lezcano commented on lib/Analysis/Allocation.cpp:
```diff
@@ -29,6 +31,30 @@ constexpr int kPtrBitWidth = 64;
+  return smem.getTotalOutDimSize() / reps;
+}
+
+static unsigned getNumScratchElemsPaddedCvt(RankedTensorType srcTy,
```
yep

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -110,6 +114,219 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+  }
+
+  LogicalResult
+  transferWithinThreadSwizzling(ConvertLayoutOp op, Value src,
```
brainfart. Fixed

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -110,6 +114,219 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+      Value offset = b.xor_(regBase, b.i32_val(regIdx));
+      auto vecAddr = b.gep(smemPtrTy, llvmElemTy, smemBase, offset,
+                           LLVM::GEPNoWrapFlags::inbounds);
+      // Lezcano: Do we want to use getFreeVariableMasks for pred or nah?
```
yep, leaving it for later

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -110,6 +114,219 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+                  Type llvmElemTy, Value smemBase,
+                  ConversionPatternRewriter &rewriter) const {
+    // Local copy of vals_ to avoid modifying the original vector
+    auto vals = vals_;
```
Changed the API to address this point and your previous one.

lezcano commented:
Added in the last two commits a couple optimisations to alleviate the arithmetic intensity of `convert_layout`s.
The first one computes addresses already in `i8`. When we computed them in `bf16`, then we ended up with many `shl 1` that were not able to be optimised by any of the compilers below.

The second optimisation was proposed by @apgoucher and it's a bit more complex to explain, but it basically boils down to the following:
We have a computation of the form `b + (a ^ c)` where `b` is the base pointer, `a` is a variable that depends on the thread and warp id, and `c` is a constant. We split `c = c1 + c2` where `c2` has the property that none of its bits are one in `a` for any thread or warp. When we have this property we can split the computation above as
```
b + (a ^ c) = b + (a ^ (c1 + c2)) = b +  (a ^ c1) + c2
```
Then, `b + (a ^ c1)` does not need to be computed for every load/store, and the `+ c2` can be passed as an immediate to LDS/STS. In short, when this optimisation applies (and it mostly applies to larger tensors, which is exactly when we want to reduce the arithmetic pressure) we generate code like:
[will fill in a minute]
